Add SPA fallback template

jeroenvisser101 opened the pull request:
This is the first draft for adding support for SPA fallback templates that can be used when generating a Nuxt project that uses the universal mode. These fallback templates will execute as if generated in `mode: 'spa'`, and are being rendered entirely on the client side.

## Pending tasks
- [x] Add test case for conflict prevention
- [x] Add docs for implementing on configuration
- [x] Add tutorial for implementation on:
    - [x] Netlify
    - [x] Firebase Hosting
    - [x] Surge

jeroenvisser101 committed:
Add SPA fallback template

pi0 commented on lib/builder/generator.js:
```diff
@@ -141,15 +142,25 @@ module.exports = class Generator {
+    if (fallback !== true && typeof fallback !== 'string') return
+
+    // True means we assume the path is 404.html (for GitHub Pages, Surge, etc.)
+    if (fallback === true) fallback = '404.html'
```
This normalize would be better moved into Options.from so that all modules will see the same option.

jeroenvisser101 replied:
Ah, totally!

pi0 commented on lib/common/options.js:
```diff
@@ -223,6 +223,7 @@ Options.defaults = {
     concurrency: 500,
     interval: 0,
     subFolders: true,
+    fallback: '200.html',
```
Why not defaulting to 404.html?

jeroenvisser101 replied:
It's to keep current behaviour default, so no BC breakage occurs.

jeroenvisser101 replied:
See https://github.com/nuxt/nuxt.js/issues/2690#issuecomment-360879305

pi0 replied:
Okay :)

pi0 reviewed (commented):
Cool 👍

jeroenvisser101 commented:
@pi0 I've incorporated the default in `Options.from` and added a test to make sure it's being transformed. Let me know if I missed anything

@Atinux I've finalized all features and added some docs that cover most of it, let me know what you think!

Atinux commented:
Love it! It's a beautiful PR you made @jeroenvisser101 

I'm merging this PR, could you tell me when the docs PR is made with Netlify example?

BTW: Thank you so much for contributing to Nuxt.js 💚

jeroenvisser101 commented:
@Atinux Thanks! Ah, I actually added it but forgot to check the box (Netlify & Github are the same, just set `generate.fallback` to true and it works without any configuration)

dotnetCarpenter commented:
@jeroenvisser101 I don't see the *Firebase Hosting* tutorial, neither in this PR or in the [documentation faq](https://nuxtjs.org/faq/github-pages). Where did you add it?

jeroenvisser101 commented:
@dotnetCarpenter sorry for the late reply, it's a bit buried in the docs, but you can find it [here](https://nuxtjs.org/guide/routing#implementation-for-firebase-hosting)
